enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,320 @@
+directly on the DNS name as it will not return any IP. Thus, the lookups will be performed on the DNS names
+which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.
+
+### Workflow Description
+
+The workflows for DNS name addition and deletion are explained in this section.
+
+#### Addition
+
+##### Regular DNS name
+* An OpenShift cluster administrator creates an EgressFirewall resource for a Namespace and adds rule(s) containing
+regular DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the ``openshift-dns`` Namespace and the Name of the CR will be same as the DNS
+name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.
+* The OVN-K master will then perform DNS lookup for each of the regular DNS names added to the EgressFirewall rules.
+* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookup for each of the
+regular DNS names.
+* As these DNS names have corresponding ``DNSName`` CRs, the ``egressfirewall`` plugin will update the ``.status`` of
+the `DNSName` CRs with the DNS name and the corresponding current IPs along with the TTL and the next lookup time based
+on the TTL. However, this update will only take place if there is a change in the exisiting IP
+addresses or next time to lookup or both for the DNS name.
+* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
+will update the ``AddressSet`` for the DNS name, which is linked with the ACL rule(s) for the corresponding EgressFirewall(s).
+* The OVN-K master will also receive the response of the DNS lookup query for the DNS name. The OVN-K master will check
+the corresponding ``DNSName`` CR's ``.status`` and if the next lookup time in the status is greater than the next lookup
+time based on the received TTL, then the corresponding CR's ``.status`` will be updated. The corresponding ``AddressSet``
+will also be updated.
+
+##### Wildcard DNS name
+* An OpenShift cluster administrator creates an EgressFirewall resource for a Namespace and adds rule(s) containing
+wildcard DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the wildcard DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard
+DNS name after replacing the ``*`` with ``wildcard`` (barring any trailinng `.`). For example, if the wildcard DNS name is
+``*.example.com``, then the Name of the corresponding CR will be ``wildcard.example.com``. This is done to adhere to the
+[Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
```

Comment:
The Name field is set to the DNS name so that there's one to one mapping with the ``DNSName`` CR and a DNS name and the CR can be fetched by the CoreDNS plugin just by using the DNS name whenever the lookup happens. I am not sure how this can be achieved without directly using the DNS name. The problem comes with wildcard DNS name.

If we have one ``DNSName`` for each EgressFirewall and the ``.status`` of ``DNSName`` containing all the details of the IPs, would it not be better to just add the information about the IPs in the ``.status`` of the EgressFirewall itself and not create a new CRD?

Reply:
> the CR can be fetched by the CoreDNS plugin just by using the DNS name whenever the lookup happens

but the plugin needs to be keeping a cache of all existing DNSName objects anyway, so it could just keep a "name → DNSName object" map in addition to that

> If we have one DNSName for each EgressFirewall and the .status of DNSName containing all the details of the IPs, would it not be better to just add the information about the IPs in the .status of the EgressFirewall itself and not create a new CRD?

yes, maybe...

Reply:
(Though that would imply that the plugin was EgressFirewall-specific and couldn't be used for other things.)

Reply:
I have already asked that wildcard dns not be a part of this e.p.  It is not in the RFE.

Reply:
I have to agree with @danwinship here in that I think having the DNS name in the CR spec rather than the CR name is a better way to go, it eliminates the complexity of wildcard substitutions since then we can have some much more explicit structure, perhaps something like below
```
spec:
  dns:
    domains:
    - a.example.com
    - b.example.com
    - c.example.com
    sudomains:
    - d.example.com // This would then be treated as *.d.example.com
```
The idea of one DNSName for each EgressFirewall then begs the question, what about one DNSName per cluster? Do we need to separate it at all if all it's doing is storing resolved addresses? Do we ever need more than one?

Reply:
Having a single `DNSName` per cluster will probably have some impact on the performance when looking up DNS names in the status of the object. 

Creating a `DNSName` for each `EgressFirewall` will introduce duplication of information as same DNS name can be used in multiple `EgressFirewalls`. So, the IPs associated with the DNS name have to be updated in all the corresponding `DNSName` objects.

I am more inclined towards having a `DNSName` object for each DNS name (regular and wildcard). The name of the `DNSName` can be a hash value and the spec will contain the corresponding DNS name. Going through the status of this `DNSName` object will have lesser performance impact.

Reply:
For example, if there are 2 EgressFirewall DNS rules, one for `www.redhat.com` and another for `*.openshift.com`, then there will be 2 `DNSName` CRs for each one of them. The  `IsRegular` and `IsWildcard` fields won't be required. Whenever, a lookup happens for `www.redhat.com`, CoreDNS will only fetch the CR corresponding to it and if required update the status. Similarly, if a lookup happens for any subdomain under `*.openshift.com`, then will only fetch the CR corresponding to it and if required update the status.

NOTE: The status of a regular DNS name will only contain one item in the `resolvedNames` field corresponding to the regular DNS name, whereas for a wildcard DNS name the `resolvedNames` will contain one item for the wildcard DNS name and additional items any other subdomains under it, if any.
```yaml
...
metadata:
  name: dns-596764f49c
  namespace: openshift-dns
spec:
  name: "www.redhat.com"
status:
  resolvedNames:
  -  dnsName: "www.redhat.com"
     info:
     - ...
```

```yaml
...
metadata:
  name: dns-747487bbfb
  namespace: openshift-dns
spec:
  name: "*.openshift.com"
status:
  resolvedNames:
  -  dnsName: "*.openshift.com"
     info:
     - ...
  -  dnsName: "docs.openshift.com"
     info:
     - ...
```

Reply:
>Having many objects subscribed will result in far more events hitting the informer than a single shared object, I think in terms of informers a single object would be more efficient.

I think only the create and delete events will be the additional ones. However, the number of update events will be the same as update will happen whenever there's a change in the IPs for a DNS name, whether we use a single object or one object each for each DNS name. The additional impact will be proportional to the number of unique DNS names used in the EgressFirewall rules.

The performance impact that I am thinking about is the time spent to traverse through the status of the CR. The `resolvedNames` field in the status is a slice and if it contains all the DNS names used in EgressFirewall rules, including those which match wildcard DNS names, then every time OVN-K will have to go through the entire slice to get the latest update in the IP addresses. It'll also have spend time to check whether IPs for each DNS name changed or not and update only for the ones that changed. If we have a CR for each DNS name, then OVN-K will immediately know, during an update event, the IPs for which DNS name changed and take appropriate actions. Same for the CoreDNS plugin when it has to update the status of the single object.

Another issue is more from the perspective of parallel updates to the single object by CoreDNS and OVN-K. It was suggested here (https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162) to have a feedback from OVN-K to CoreDNS. If we have a single object, parallel updates may cause some race condition problems with out of order updates. Additionally, CoreDNS pods run as DaemonSets and each pod runs independent of the other. Thus, more than one pod may receive DNS lookup requests for different DNS names simultaneously. When updating the status of the single object chances are there that one update may be done over the other one and some updates may get lost because of this.

> Is CoreDNS fetching the CR directly? Will it be using an informer?

I was planning to have both an informer for the create and delete events in the CoreDNS plugin. But when updating the object, it would be first fetched then the status would be checked with the current IPs and lookup time information. If it requires an update then update the status. OVN-K will be creating and deleting the CRs, so it will be using an informer for the update event to get the changes in the IP information of a DNS name.

Reply:
> The performance impact that I am thinking about is the time spent to traverse through the status of the CR. The `resolvedNames` field in the status is a slice and if it contains all the DNS names used in EgressFirewall rules, including those which match wildcard DNS names, then every time OVN-K will have to go through the entire slice to get the latest update in the IP addresses. It'll also have spend time to check whether IPs for each DNS name changed or not and update only for the ones that changed. If we have a CR for each DNS name, then OVN-K will immediately know, during an update event, the IPs for which DNS name changed and take appropriate actions. Same for the CoreDNS plugin when it has to update the status of the single object.

This makes sense to me.  Keep in mind that there is potential fanout:  An EgressFirewall can have rules with many DNS names or wildcards, and a wildcard can result in an EgressFirewallDNSName with many values under `status.resolvedNames`.  

> Another issue is more from the perspective of parallel updates to the single object by CoreDNS and OVN-K. It was suggested here ([#1335 (comment)](https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162)) to have a feedback from OVN-K to CoreDNS. If we have a single object, parallel updates may cause some race condition problems with out of order updates. Additionally, CoreDNS pods run as DaemonSets and each pod runs independent of the other. Thus, more than one pod may receive DNS lookup requests for different DNS names simultaneously. When updating the status of the single object chances are there that one update may be done over the other one and some updates may get lost because of this.

Agreed.  We will have many CoreDNS pods updating the EgressFirewallDNSName objects to update `status.resolvedNames` (by default, a cluster has 1 DNS pod per node), and we don't want these pods to be constantly in conflict with each other, which would cause frequent retries and unnecessarily overhead for both the CoreDNS plugin and kube-apiserver.